sgx/common/src/retrieve_key_protocol.c

```diff
@@ -0,0 +1,1164 @@
+/**
+ * @file retrieve_key_protocol.c
+ *
+ * @brief Provides implementation for functionality directly supporting
+ *        creating, validating, and parsing the kmyth 'retrieve key'
+ *        protocol messages.
+ */
+
+#include "retrieve_key_protocol.h"
+
+/*****************************************************************************
+ * extract_identity_bytes_from_x509()
+ ****************************************************************************/
+int extract_identity_bytes_from_x509(X509 *cert_in, X509_NAME **identity_out)
+{
+  // extract 'subject name' from input certificate
+  //   Note: The returned X509_NAME is an internal pointer
+  //         that should NOT be freed.
+  X509_NAME *subj_name = X509_get_subject_name(cert_in);
+  if (subj_name == NULL)
+  {
+    kmyth_sgx_log(LOG_ERR, "extraction of certificate's subject name failed");
+    return EXIT_FAILURE;
+  }
+
+  *identity_out = X509_NAME_dup(subj_name);
+  if (identity_out == NULL)
+  {
+    kmyth_sgx_log(LOG_ERR, "duplication of X509_NAME struct failed");
+    return EXIT_FAILURE;
+  }
+
+  return EXIT_SUCCESS;
+}
+
+/*****************************************************************************
+ * append_msg_signature()
+ ****************************************************************************/
+int append_msg_signature(EVP_PKEY * sign_key,
+                         ECDHMessage * msg)
+{
+  // compute message signature
+  unsigned char *signature_bytes = NULL;
+  int signature_len = 0;
+
+  if (EXIT_SUCCESS != ec_sign_buffer(sign_key,
+                                     msg->body,
+                                     msg->hdr.msg_size,
+                                     &signature_bytes,
+                                     &signature_len))
+  {
+    kmyth_sgx_log(LOG_ERR, "error signing buffer");
+    free(signature_bytes);
```

Comment:
Attempted to address in commit [d46480a](https://github.com/NationalSecurityAgency/kmyth/pull/168/commits/d46480a9ddd27de6f27753b32bddf6130dcb8050).

Removed the `free(signature_bytes)` line in the error condition following the call to `ec_sign_buffer()`.

In the ec_sign_buffer() function, added a line to free the signature bytes buffer on a failed call to EVP_SignFinal().

While not part of the comment, I also changed this ec_sign_buffer() code to use `calloc()` rather than `OPENSSL_malloc()` to allocate this buffer, as we are not allocating an OPENSSL specific object here.